TTS/tts/models/naturalspeech2.py

```diff
@@ -0,0 +1,1223 @@
+
+    def _freeze_layers(self):
+        if self.args.freeze_phoneme_encoder:
+            for param in self.phoneme_encoder.parameters():
+                param.requires_grad = False
+
+        if self.args.freeze_prompt_encoder:
+            for param in self.prompt_encoder.parameters():
+                param.requires_grad = False
+
+        if self.args.freeze_duration_predictor:
+            for param in self.duration_predictor.parameters():
+                param.requires_grad = False
+
+        if self.args.freeze_pitch_predictor:
+            for param in self.pitch_predictor.parameters():
+                param.requires_grad = False
+
+        if self.args.freeze_diffusion:
+            for param in self.diffusion.parameters():
+                param.requires_grad = False
+
+    def _forward_aligner(
+        self, x: torch.FloatTensor, y: torch.FloatTensor, x_mask: torch.IntTensor, y_mask: torch.IntTensor
+    ) -> Tuple[torch.IntTensor, torch.FloatTensor, torch.FloatTensor, torch.FloatTensor]:
+        """Aligner forward pass.
+
+        1. Compute a mask to apply to the attention map.
+        2. Run the alignment network.
+        3. Apply MAS to compute the hard alignment map.
+        4. Compute the durations from the hard alignment map.
+
+        Args:
+            x (torch.FloatTensor): Input sequence.
+            y (torch.FloatTensor): Output sequence.
+            x_mask (torch.IntTensor): Input sequence mask.
+            y_mask (torch.IntTensor): Output sequence mask.
+
+        Returns:
+            Tuple[torch.IntTensor, torch.FloatTensor, torch.FloatTensor, torch.FloatTensor]:
+                Durations from the hard alignment map, soft alignment potentials, log scale alignment potentials,
+                hard alignment map.
+
+        Shapes:
+            - x: :math:`[B, T_en, C_en]`
+            - y: :math:`[B, T_de, C_de]`
+            - x_mask: :math:`[B, 1, T_en]`
+            - y_mask: :math:`[B, 1, T_de]`
+
+            - alignment_hard: :math:`[B, T_en]`
+            - alignment_soft: :math:`[B, T_en, T_de]`
+            - alignment_logprob: :math:`[B, 1, T_de, T_en]`
+            - alignment_mas: :math:`[B, T_en, T_de]`
+        """
+        attn_mask = torch.unsqueeze(x_mask, -1) * torch.unsqueeze(y_mask, 2)
+        alignment_soft, alignment_logprob = self.aligner(y.transpose(1, 2), x, x_mask, None)
+        assert not torch.isnan(alignment_soft).any()
+
+        alignment_mas = maximum_path(
+            alignment_soft.squeeze(1).transpose(1, 2).contiguous(), attn_mask.squeeze(1).contiguous()
+        )
+        alignment_hard = torch.sum(alignment_mas, -1).int()
+        alignment_soft = alignment_soft.squeeze(1).transpose(1, 2)
+        return alignment_hard, alignment_soft, alignment_logprob, alignment_mas
+
+    @staticmethod
+    def _set_cond_input(aux_input: Dict):
+        # [TODO] use this
+        return None
+
+    def forward(  # pylint: disable=dangerous-default-value
```

Comment:
looks like the first Q-K-V Attention is missing that is with random m embeddings in Fig3 in the paper

Reply:
i have added this in diffusion.py

Reply:
class Denoise

Reply:
then no worries 👍